### PR TITLE
Make the demangler in the runtime use stack allocated memory.

### PR DESCRIPTION
**Pull request by eeckstein**
A big part of the change is to reduce the size of demangler Node: This is done by disallowing nodes with children to also have index or text payloads.
In some cases those payloads were not needed anyway, because the information can be derived later.
In other cases the fix was to insert an additional child node with the index/text payload.

The demangler now supports stack allocated memory for its allocator.
The demangler can be initialized with a preallocated memory on the stack. Only in case of an overflow, the bump pointer allocator mallocs new memory.
Also, support that a new instance of a demangler can "borrow" the free memory from an existing demangler. This is useful because in the runtime the demangler is invoked recursively. With this feature, all the nested demanglers can share a single stack allocated space.

rdar://problem/47357709

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci test

**Comment by eeckstein** [comment]
@swift-ci benchmark

File: lib/Demangling/Demangler.cpp [hunk 1]
```diff
@@ -319,16 +402,18 @@ void Node::reverseChildren(size_t StartingAt) {
...
     slab = prev;
   }
 }
   
 void NodeFactory::clear() {
+  assert(!isBorrowed);
   if (CurrentSlab) {
+#ifdef NODE_FACTORY_DEBUGGING
+    std::cerr << indent() << "## clear: allocated memory = " << allocatedMemory  << "\n";
+#endif
```
Comment (jckarter) [inline lib/Demangling/Demangler.cpp]: Would LLVM's DEBUG(...) macros work from the runtime?

Comment (eeckstein) [reply lib/Demangling/Demangler.cpp]: I didn't try, but we are not linking llvm to the runtime

**Review by jckarter — APPROVED** [review]
Looks good, thanks for doing this Erik!

**Pull request merged**